pkg/controller/ground/bootstrap/dns/coredns.go

```diff
@@ -0,0 +1,310 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "system:coredns",
+			Labels: map[string]string{
+				"kubernetes.io/bootstrapping":     "rbac-defaults",
+				"addonmanager.kubernetes.io/mode": "EnsureExists",
+			},
+			Annotations: map[string]string{
+				"rbac.authorization.kubernetes.io/autoupdate": "true",
+			},
+		},
+		RoleRef: rbac.RoleRef{
+			APIGroup: rbac.GroupName,
+			Kind:     "ClusterRole",
+			Name:     "system:coredns",
+		},
+		Subjects: []rbac.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      "coredns",
+				Namespace: metav1.NamespaceSystem,
+			},
+		},
+	})
+}
+
+func createCoreDNSConfigMap(client clientset.Interface, domain string) error {
+	if domain == "" {
+		domain = spec.MustDefaultString("KlusterSpec", "dnsDomain")
+	}
+
+	manifest := `
+.:53 {
+    errors
+    health
+    kubernetes {{ .Domain }} in-addr.arpa ip6.arpa {
+        pods insecure
+        upstream
+        fallthrough in-addr.arpa ip6.arpa
+        ttl 30
+    }
+    prometheus :9153
+    forward . /etc/resolv.conf
+    cache 30
+    loop
+    reload
+    loadbalance
+}`
+
+	data := struct{ Domain string }{domain}
+
+	template, err := bootstrap.RenderManifest(manifest, data)
+	if err != nil {
+		return err
+	}
+
+	return bootstrap.CreateOrUpdateConfigMap(client, &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "coredns",
+			Namespace: metav1.NamespaceSystem,
+			Labels: map[string]string{
+				"addonmanager.kubernetes.io/mode": "EnsureExists",
+			},
+		},
+		Data: map[string]string{
+			"Corefile": string(template),
+		},
+	})
+}
+
+func createCoreDNSService(client clientset.Interface, clusterIP string) error {
+	if clusterIP == "" {
+		return errors.New("ClusterIP is missing")
+	}
+
+	return bootstrap.CreateOrUpdateService(client, &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kube-dns",
+			Namespace: metav1.NamespaceSystem,
+			Labels: map[string]string{
+				"k8s-app":                         "kube-dns",
+				"kubernetes.io/cluster-service":   "true",
+				"addonmanager.kubernetes.io/mode": "Reconcile",
+				"kubernetes.io/name":              "CoreDNS",
+			},
+			Annotations: map[string]string{
+				"prometheus.io/port":   "9153",
+				"prometheus.io/scrape": "true",
+			},
+			ResourceVersion: "0",
+		},
+		Spec: v1.ServiceSpec{
+			Selector: map[string]string{
+				"k8s-app": "kube-dns",
+			},
+			ClusterIP: clusterIP,
+			Ports: []v1.ServicePort{
+				{Name: "dns", Port: 53, Protocol: v1.ProtocolUDP},
+				{Name: "dns-tcp", Port: 53, Protocol: v1.ProtocolTCP},
+				{Name: "metrics", Port: 9153, Protocol: v1.ProtocolTCP},
+			},
+		},
+	})
+}
+
+func createCoreDNSDeployment(client clientset.Interface, repository, version string) error {
+	if repository == "" {
+		repository = "sapcc"
+	}
+
+	if version == "" {
+		version = "1.3.1"
+	}
+
+	manifest := `
+apiVersion: extensions/v1beta1
+kind: Deployment
+metadata:
+  name: coredns
+  namespace: kube-system
+  labels:
+    k8s-app: kube-dns
```

Comment:
Is the use of `kube-dns` in this deployment deliberate? Like meaning "the Kubernetes dns service" of is this a copy and paste error from the old `kube-dns` stack?

Reply:
Nicely spotted. I saw it as well and wondered. It is aligned with the default deployment here: https://github.com/kubernetes/kubernetes/blob/master/cluster/addons/dns/coredns/coredns.yaml.base#L87

My assumption is, that it relates to the `kube-dns` service. Maybe to smooth out an upgrade without interruption.